File: edukate-common/src/main/java/io/github/sanyavertolet/edukate/common/notifications/InviteNotificationCreateRequest.java
package io.github.sanyavertolet.edukate.common.notifications;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonTypeName;
import jakarta.validation.constraints.NotBlank;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.SuperBuilder;

@Getter
@Setter
@SuperBuilder
@ToString(callSuper = true)
@JsonTypeName("invite")
public final class InviteNotificationCreateRequest extends BaseNotificationCreateRequest {
    @NotBlank
    private String inviterName;
    @NotBlank
    private String bundleName;
    @NotBlank
    private String bundleShareCode;

    @JsonCreator
    public InviteNotificationCreateRequest(
            String uuid, String targetUserId, String inviterName, String bundleName, String bundleShareCode
    ) {
        super(uuid, targetUserId);
        this.inviterName = inviterName;
        this.bundleName = bundleName;
        this.bundleShareCode = bundleShareCode;
    }
}

File: edukate-common/src/main/kotlin/io/github/sanyavertolet/edukate/common/notifications/InviteNotificationCreateRequest.kt
package io.github.sanyavertolet.edukate.common.notifications

import com.fasterxml.jackson.annotation.JsonTypeName
import jakarta.validation.constraints.NotBlank
import java.util.UUID

@JsonTypeName("invite")
data class InviteNotificationCreateRequest(
    override val uuid: String = UUID.randomUUID().toString(),
    override val targetUserId: Long,
    @field:NotBlank val inviterName: String,
    @field:NotBlank val problemSetName: String,
    @field:NotBlank val problemSetShareCode: String,
) : BaseNotificationCreateRequest {
    companion object {
        @JvmStatic
        fun from(targetUserId: Long, inviterName: String, problemSetName: String, problemSetShareCode: String) =
            InviteNotificationCreateRequest(
                targetUserId = targetUserId,
                inviterName = inviterName,
                problemSetName = problemSetName,
                problemSetShareCode = problemSetShareCode,
            )
    }
}
